Sources/Datadog/Core/System/AppStateListener.swift

```diff
@@ -0,0 +1,150 @@
+            }
+            if event.isActive {
+                lastActiveStartDate = event.date
+            } else {
+                lastActiveStartDate = nil
+            }
+        }
+        return duration
+    }
+
+    var didRunInBackground: Bool {
+        return !initialState.isActive || !finalState.isActive
+    }
+
+    private func isActive(at date: Date) -> Bool {
+        if date <= initialState.date {
+            // we assume there was no change before initial state
+            return initialState.isActive
+        } else if finalState.date <= date {
+            // and no change after final state
+            return finalState.isActive
+        }
+        var active = initialState.isActive
+        for change in changes {
+            if date < change.date {
+                break
+            }
+            active = change.isActive
+        }
+        return active
+    }
+}
+
+internal protocol AppStateListening: class {
+    var history: AppStateHistory { get }
+}
+
+internal class AppStateListener: AppStateListening {
+    typealias Snapshot = AppStateHistory.Snapshot
+
+    static var shared = AppStateListener()
+
+    private let dateProvider: DateProvider
+    private var _appStateHistory: AppStateHistory
+    var history: AppStateHistory {
+        objc_sync_enter(self)
+        defer { objc_sync_exit(self) }
+        _appStateHistory.finalDate = dateProvider.currentDate()
+        return _appStateHistory
+    }
+
+    private static var isAppActive: Bool {
+        if Thread.isMainThread {
+            return UIApplication.managedShared?.applicationState == .active
+        } else {
+            return DispatchQueue.main.sync {
+                UIApplication.managedShared?.applicationState == .active
+            }
+        }
+    }
+
+    init(dateProvider: DateProvider = SystemDateProvider()) {
+        self.dateProvider = dateProvider
+
+        let isAppActive = AppStateListener.isAppActive
+        let currentState = Snapshot(isActive: isAppActive, date: dateProvider.currentDate())
+        _appStateHistory = AppStateHistory(
+            initialState: currentState,
+            finalDate: currentState.date
+        )
+        let nc = NotificationCenter.default
+        nc.addObserver(self, selector: #selector(appWillResignActive), name: UIApplication.willResignActiveNotification, object: nil)
+        nc.addObserver(self, selector: #selector(appDidBecomeActive), name: UIApplication.didBecomeActiveNotification, object: nil)
+    }
+
+    @objc
+    private func appWillResignActive() {
+        objc_sync_enter(self)
+        _appStateHistory.add(change: Snapshot(isActive: false, date: dateProvider.currentDate()))
+        objc_sync_exit(self)
+    }
+    @objc
+    private func appDidBecomeActive() {
+        objc_sync_enter(self)
+        _appStateHistory.add(change: Snapshot(isActive: true, date: dateProvider.currentDate()))
+        objc_sync_exit(self)
```

Comment:
This will block the main thread (the _writer thread_) if any other thread (any _reader thread_) is accessing the `_appStateHistory`. If there are 100 simultaneous reads, this write operation will block the app UI until all of them complete 😞.